src/treelearner/feature_histogram.hpp

```diff
@@ -80,44 +85,61 @@ class FeatureHistogram {
+    could_be_splittable_ = false;
     double gain_shift = GetLeafSplitGain(sum_gradient, sum_hessian,
                                          meta_->config->lambda_l1, meta_->config->lambda_l2, meta_->config->max_delta_step);
     double min_gain_shift = gain_shift + meta_->config->min_gain_to_split;
+
+    // at this point, the following arrays contain the constraints applied on every part of the leaf
+    // since we are splitting the leaf in 2, we can compute the cumulative / minimum maximum in both directions
+    constraints.ComputeCumulativeExtremums();
+
     if (meta_->num_bin > 2 && meta_->missing_type != MissingType::None) {
       if (meta_->missing_type == MissingType::Zero) {
-        FindBestThresholdSequence(sum_gradient, sum_hessian, num_data, min_constraint, max_constraint, min_gain_shift, output, -1, true, false);
-        FindBestThresholdSequence(sum_gradient, sum_hessian, num_data, min_constraint, max_constraint, min_gain_shift, output, 1, true, false);
+        FindBestThresholdSequence(
+            sum_gradient, sum_hessian, num_data, min_gain_shift, output, -1,
+            true, false, constraints);
+        FindBestThresholdSequence(
+            sum_gradient, sum_hessian, num_data, min_gain_shift, output, 1,
+            true, false, constraints);
       } else {
-        FindBestThresholdSequence(sum_gradient, sum_hessian, num_data, min_constraint, max_constraint, min_gain_shift, output, -1, false, true);
-        FindBestThresholdSequence(sum_gradient, sum_hessian, num_data, min_constraint, max_constraint, min_gain_shift, output, 1, false, true);
+        FindBestThresholdSequence(
+            sum_gradient, sum_hessian, num_data, min_gain_shift, output, -1,
+            false, true, constraints);
+        FindBestThresholdSequence(
+            sum_gradient, sum_hessian, num_data, min_gain_shift, output, 1,
+            false, true, constraints);
       }
     } else {
-      FindBestThresholdSequence(sum_gradient, sum_hessian, num_data, min_constraint, max_constraint, min_gain_shift, output, -1, false, false);
+      FindBestThresholdSequence(
+          sum_gradient, sum_hessian, num_data, min_gain_shift, output, -1,
+          false, false, constraints);
       // fix the direction error when only have 2 bins
       if (meta_->missing_type == MissingType::NaN) {
         output->default_left = false;
       }
     }
     output->gain -= min_gain_shift;
     output->monotone_type = meta_->monotone_type;
-    output->min_constraint = min_constraint;
-    output->max_constraint = max_constraint;
   }
 
-  void FindBestThresholdCategorical(double sum_gradient, double sum_hessian, data_size_t num_data,
-                                    double min_constraint, double max_constraint,
-                                    SplitInfo* output) {
+  void FindBestThresholdCategorical(
+      double sum_gradient, double sum_hessian, data_size_t num_data,
+      SplitInfo *output, SplittingConstraints& constraints) {
```

Comment:
I think categorcial features don't need the constraints, for there are not numerical ordering relations.

Reply:
There are 3 things I want to highlight regarding this function. 

First from a code point of view, in the same file, the function find_best_threshold_fun_ is a bind for both FindBestThresholdCategorical and FindBestThresholdNumerical so these 2 functions need to have the same template. So I would have to change a significant amount of code to make the 2 not have the same template.

Second, in theory, Say we have a dataset with 2 features, 1 being categorical and 1 numerical, and we apply a monotonic constraint on the numerical one. Then even when when we split the data with respect to the categorical feature, then the best constraints for the resulting leaves may vary depending on which categories you split the categorical feature on. For example if you look at example 11 of the pdf report I posted in my initial comment, then you can imagine that the unconstrained feature is actually a categorical feature, and you see that the constraints applied to the leaves also depend on the categorical splits. Currently this is not implemented in the code, and I don't plan on implementing it in this PR, because it has a pretty marginal use. But I or someone else may, one day, want to implement that and the current way the code works is perfect for that.

Finally, there are some research papers that want to apply monotone constraints to categorical data (e.g. credit ratings, yes-no questions...). So maybe one day someone will want to make a PR implementing monotone constraints for categorical features.

For these reasons, I don't agree with removing the constraints argument from the categorical features splitting function you mentioned. But if you still think that's best in light of the above then I will do it.

Reply:
yeah, you can keep these function arguments. Actually, we will have some upgrade for the categorical features (refer to #2463), and the old categorical split will be abandoned.